src/js/reducers/datasetsReducer.js

```diff
@@ -38,6 +39,41 @@ function datasetsReducer(state, action) {
     }));
   }
 
+  if (action.type === ACTIONS.FILTER_DATASET) {
+    state = setIn(state, action.id + '._filter',
+    Immutable.fromJS({
+      expression: action.expression
+    }));
+  }
+
+  var transforms = getIn(state, action.id + '._transforms');
+  if (action.type === ACTIONS.ADD_DATA_TRANSFORM) {
+
+    if (!transforms) {
+      transforms = [];
+    }
+    var result = deepEquality.indexOf(action.transformSpec, transforms);
+    if (result === -1) {
+      // if transform doesn't exist already
+      transforms.push(action.transformSpec);
+    }
+
+    state = setIn(state, action.id + '._transforms', transforms);
+  }
+
+  if (action.type === ACTIONS.EDIT_DATA_TRANSFORM) {
+
+    if (!transforms) {
+      transforms = [];
+    }
+    var index = action.specId;
+    if (index > -1) {
+      // if transform exist already
+      transforms[index] = action.newSpec;
+    }
+    state = setIn(state, action.id + '._transforms', transforms);
```

Comment:
When editing, you can probably call `setIn` using the passed in index. For example, `return setIn(state, action.id + '.transforms.' + index, Immutable.fromJS(action.newSpec))`.